src/WebJobs.Script/Workers/WorkerConcurrencyManager.cs

```diff
@@ -0,0 +1,225 @@
+                            Interval = _workerConcurrencyOptions.CheckInterval.TotalMilliseconds,
+                        };
+
+                        _timer.Elapsed += OnTimer;
+                        _timer.Start();
+                    }
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            if (_timer != null)
+            {
+                _logger.LogDebug("Stopping dynamic worker concurrency monitoring.");
+                _timer.Stop();
+            }
+            return Task.CompletedTask;
+        }
+
+        internal async void OnTimer(object sender, System.Timers.ElapsedEventArgs e)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            try
+            {
+                var workerStatuses = await _functionInvocationDispatcher.GetWorkerStatusesAsync();
+
+                if (NewWorkerIsRequired(workerStatuses, _addWorkerStopwatch.Elapsed))
+                {
+                    await _functionInvocationDispatcher.StartWorkerChannel();
+                    _logger.LogDebug("New worker is added.");
+                    _addWorkerStopwatch.Restart();
+                }
+            }
+            catch (Exception ex)
+            {
+                // don't allow background exceptions to escape
+                _logger.LogError(ex.ToString());
+            }
+            _timer.Start();
+        }
+
+        internal bool NewWorkerIsRequired(IDictionary<string, WorkerStatus> workerStatuses, TimeSpan timeSinceLastNewWorker)
+        {
+            if (timeSinceLastNewWorker < _workerConcurrencyOptions.AdjustmentPeriod)
+            {
+                return false;
+            }
+
+            bool result = false;
+            if (workerStatuses.All(x => x.Value.IsReady))
+            {
+                // Check how many channels are overloaded
+                List<WorkerStatusDetails> descriptions = new List<WorkerStatusDetails>();
+                foreach (string key in workerStatuses.Keys)
+                {
+                    WorkerStatus workerStatus = workerStatuses[key];
+                    bool overloaded = IsOverloaded(workerStatus);
+                    descriptions.Add(new WorkerStatusDetails()
+                    {
+                        WorkerId = key,
+                        WorkerStatus = workerStatus,
+                        IsOverloaded = overloaded
+                    });
+                }
+
+                int overloadedCount = descriptions.Where(x => x.IsOverloaded == true).Count();
+                if (overloadedCount > 0)
+                {
+                    if (workerStatuses.Count() < _workerConcurrencyOptions.MaxWorkerCount)
+                    {
+                        _logger.LogDebug($"Adding a new worker, overloaded workers = {overloadedCount}, initialized workers = {workerStatuses.Count()} ");
```

Comment:
The caller is already logging. I think this log should be removed, and update the other log if needed. This method doesn't add the worker, it just checks if one needs to be added